plat/kvm: Add Multiboot2 support

mariapana opened the pull request:
This pull request introduces support for the Multiboot2 protocol, a successor to the original Multiboot variant.

<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): x86_64
 - Platform(s): kvm
 - Application(s): N/A


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

Since QEMU currently doesn't natively support the Multiboot2 bootloader format, options like `-kernel` are not directly applicable. To test with Multiboot2 images, you'll need to create `.iso` files using the `mkukimg` script. For a detailed example of how to create and boot a Multiboot2 ISO image using QEMU, refer to this [`nginx` sample](https://gist.github.com/mariapana/90ab2e61715e812def0c7582c9482626).



### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->
* Add Multiboot2 header support to ELF64 binaries
* Integrate Multiboot2 support in KVM platform by adding `multiboot2.h`, `multiboot2.c` and modifying `multiboot.S`
* Support building Multiboot2 images by updating the `mkukimg` script
* Remove `ELF64_TO_32` for the Multiboot protocol

mogasergiu commented on plat/kvm/Config.uk:
```diff
@@ -35,6 +35,12 @@ config KVM_BOOT_PROTO_MULTIBOOT
 
+config KVM_BOOT_PROTO_MULTIBOOT2
+	bool "Multiboot2"
+	depends on KVM_VMM_QEMU && !KVM_VMM_FIRECRACKER && ARCH_X86_64
```
Drop ` KVM_VMM_QEMU && !KVM_VMM_FIRECRACKER`, it is a generic x86 boot protocol, platform agnostic.

mogasergiu commented on plat/kvm/Linker.uk:
```diff
@@ -3,6 +3,9 @@ ifeq ($(CONFIG_KVM_BOOT_PROTO_MULTIBOOT),y)
 KVM_LDFLAGS-y += -Wl,-m,elf_x86_64
 KVM_LDFLAGS-y += -Wl,--entry=_multiboot_entry
 ELF64_TO_32 = y
+else ifeq ($(CONFIG_KVM_BOOT_PROTO_MULTIBOOT2),y)
```
This looks like it could have been a simple `||` instead of an `else if`.
```C
ifeq ($(findstring y,$(CONFIG_KVM_BOOT_PROTO_MULTIBOOT) $(CONFIG_KVM_BOOT_PROTO_MULTIBOOT2)),y)
```

mogasergiu commented on plat/kvm/Linker.uk:
```diff
@@ -81,6 +84,9 @@ $(KVM_IMAGE): $(KVM_IMAGE).dbg
 ifeq ($(ELF64_TO_32),y)
```
Ughh... this looks so ugly and inconsistent now... I think I am going to ask you to replace this with a `ifeq ($(CONFIG_KVM_BOOT_PROTO_MULTIBOOT),y)`... I know this is out of this PR's scope but at the time when I first added this I thought it would end up being used by other protocols as well... yeah :sweat_smile:

mogasergiu commented on plat/kvm/include/kvm-x86/multiboot2.h:
```diff
@@ -0,0 +1,381 @@
+ * SOFTWARE.
+ */
+
+#ifndef MULTIBOOT_HEADER
```
Mimic out multiboot header as well, use guards like these
```C
#ifndef __MULTIBOOT2_H__
#define __MULTIBOOT2_H__
```

mogasergiu commented on plat/kvm/include/kvm-x86/multiboot2.h:
```diff
@@ -0,0 +1,381 @@
+#define MULTIBOOT_HEADER 1
+
+/*  How many bytes from the start of the file we search for the header. */
+#define MULTIBOOT_SEARCH 32768
```
Can you restyle all of the macro's here so that their definition is separated by several tabs, similar to `multiboot.h`. At the moment it is just using one whitespace.

mogasergiu commented on plat/kvm/x86/multiboot.S:
```diff
@@ -38,7 +38,11 @@
 #include <uk/asm.h>
 #include <uk/reloc.h>
 
+#ifdef CONFIG_KVM_BOOT_PROTO_MULTIBOOT
```
We tend to use `#if` instead of `#ifdef` for boolean configurations. So I expect this to be like
```C
#if CONFIG_KVM_BOOT_PROTO_MULTIBOOT
#include <kvm-x86/multiboot.h>
#elif  CONFIG_KVM_BOOT_PROTO_MULTIBOOT2
#include <kvm-x86/multiboot2.h>
#else /* !CONFIG_KVM_BOOT_PROTO_MULTIBOOT && !CONFIG_KVM_BOOT_PROTO_MULTIBOOT2 */
#error "Unknown multiboot version"
#endif
```

mogasergiu commented on plat/kvm/x86/multiboot.S:
```diff
@@ -48,9 +52,15 @@
 lcpu_bootstack:
 
 .section .rodata
+#ifdef CONFIG_KVM_BOOT_PROTO_MULTIBOOT
```
Ditto, but also, notice how you only need to have the entry symbol defined differently, not the boostack or the startup args symbols
```C
lcpu_boot_startup_args:
#if CONFIG_KVM_BOOT_PROTO_MULTIBOOT
	ur_data	quad, multiboot_entry, 8
#elif  CONFIG_KVM_BOOT_PROTO_MULTIBOOT2
	ur_data	quad, multiboot_entry2, 8
#else /* !CONFIG_KVM_BOOT_PROTO_MULTIBOOT && !CONFIG_KVM_BOOT_PROTO_MULTIBOOT2 */
#error "Unknown multiboot version"
#endif
	ur_data	quad, lcpu_bootstack, 8
```

mogasergiu commented on plat/kvm/x86/multiboot.S:
```diff
@@ -63,8 +73,11 @@ lcpu_boot_startup_args:
 ENTRY(_multiboot_entry)
-	/* only multiboot is supported for now */
-	cmpl	$MULTIBOOT_BOOTLOADER_MAGIC, %eax
+	#ifdef CONFIG_KVM_BOOT_PROTO_MULTIBOOT
```
Ditto let's define a wrapper macro somewhere in the beginning here like so and use that
```C
#if CONFIG_KVM_BOOT_PROTO_MULTIBOOT
#define MULTIBOOT_MAGIC MULTIBOOT_BOOTLOADER_MAGIC
#elif  CONFIG_KVM_BOOT_PROTO_MULTIBOOT2
#define MULTIBOOT_MAGIC MULTIBOOT_BOOTLOADER_MAGIC2
#else /* !CONFIG_KVM_BOOT_PROTO_MULTIBOOT && !CONFIG_KVM_BOOT_PROTO_MULTIBOOT2 */
#error "Unknown multiboot version"
#endif
```
Also, do not indent `#ifdef`'s or `#define`'s

mogasergiu commented on plat/kvm/x86/multiboot2.c:
```diff
@@ -0,0 +1,208 @@
+	__sz offset;
+
+	__sz mmap_entry_cnt = (mmap_tag->size - sizeof(struct multiboot_tag_mmap)) /
+			    		 mmap_tag->entry_size;
```
Misaligned with opened paranthesis from the line above (which goes past 80 characters).

mogasergiu commented on plat/kvm/x86/multiboot2.c:
```diff
@@ -0,0 +1,208 @@
+			    		 mmap_tag->entry_size;
+
+	for (offset = 0; offset < mmap_entry_cnt;
+		offset++) {
```
This is also misaligned but it will be unproductive if I keep spamming for each single coding style issue. I would advise using `support/scripts/checkpatch.pl --strict` on your patches and solve the warnings/errors.

mogasergiu commented on plat/kvm/x86/multiboot2.c:
```diff
@@ -0,0 +1,208 @@
+	for (struct multiboot_tag *tag = (struct multiboot_tag *)(addr + 8);
+	     tag->type != MULTIBOOT_TAG_TYPE_END;
+	     tag = (struct multiboot_tag *)((multiboot_uint8_t *)tag
+					    + ((tag->size + 7) & ~7))) {
```
I will want some comment or macro surrounding this as it does not tell me much about this formula.

mogasergiu commented on support/scripts/multiboot2.py:
```diff
@@ -0,0 +1,266 @@
+
+def add_tags(tags_opt):
+    tags_hdr = b""
+    for tag_type in tags_opt:
```
Nice! But instead of `elif`, let's just have an `else` that throws an exception for unknown tag.

mogasergiu reviewed (changes requested):
Ok left some comments regarding some styling issues. There are other places I could've left comments at but it would have been unproductive as these can simply be solved by a `checkpatch.pl --strict <your patch files here>`.

So for now I will ask that you update accordingly with these and have the following commit history schema:
- first commit adds the script
- second commit adds the rest

So at the end you should only have two commits for now. The relocation tag and whatever we implement later will come separately.

mariapana committed:
support/scripts: Add Multiboot2 header support to ELF64 binaries

This script manipulares ELF64 binaries for compatibility with the
Multiboot2 protocol. The script prepends a Multiboot2 header, along with
an updated ELF header to manage updated offsets effectively.

The final image will have, in this order:
- ELF64 header
- Multiboot2 header
- ELF64 Program headers
- Original ELF64 binary
- ELF64 Section headers

Signed-off-by: Maria Pana <maria.pana4@gmail.com>

mariapana committed:
plat/kvm: Integrate Multiboot2 support in KVM platform

This commit introduces support for Multiboot2 in the KVM platform.

- Makefile, Linker and Configuration Files: updated to include Multiboot2
dependencies and settings.
- C Source Files: introduced `multiboot2.h` following the GNU
specification and `multiboot2.c` which processes the data from the tags
and manages memory regions accordingly.
- Assembly Source File: updated `multiboot.S` to support both Multiboot
and Multiboot2 protocols.

Checkpatch-Ignore: AVOID_EXTERNS
Checkpatch-Ignore: STRNCPY
Checkpatch-Ignore: STRCPY

Signed-off-by: Maria Pana <maria.pana4@gmail.com>

mariapana committed:
support/scripts: Support building Multiboot2 images

This commit updates the `mkukimg` script to include support for creating
bootable images using GRUB with Multiboot2.

The script now offers options for configuring GRUB to load kernels with
either Multiboot or Multiboot2, done by splitting the previous `grub`
bootloader option into `grubmb` and `grubmb2`, each calling their
correspondent functions (`mkukmbiso` or `mkukmb2iso`).

Signed-off-by: Maria Pana <maria.pana4@gmail.com>

mogasergiu commented on support/scripts/multiboot2.py:
```diff
@@ -0,0 +1,245 @@
+    tags_hdr = b""
+    for tag_type in tags_opt:
+        if (
+            tag_type == MULTIBOOT_HEADER_TAG_END
```
uhh, this `if` does not look that pythonesque to me :sweat_smile:

mogasergiu commented on support/scripts/multiboot2.py:
```diff
@@ -0,0 +1,245 @@
+            tag_type == MULTIBOOT_HEADER_TAG_END
+            or tag_type == MULTIBOOT_HEADER_TAG_MODULE_ALIGN
+        ):
+            size = 8
```
what is this `size` and does it need to be assigned in each iteration? seems pretty constant to me.

mariapana replied:
The `size` field is the total size the tag will have with all of its fields. In the case of the implemented tags, it is indeed equal to 8. However, there are other tags whose sizes can differ, as stated in the specification  (e.g. `MULTIBOOT_HEADER_TAG_RELOCATABLE`  would have `size` equal to 24). Should I use a macro for now and change it up once more tags are added?

mogasergiu replied:
Understood. There can only be one instance of a tag type correct? I am thinking maybe we should just create an individual function for each. E.g.
```python
def get_multiboot2_hdr():
    mb2_magic = MULTIBOOT_HEADER_MAGIC
    mb2_architecture = MULTIBOOT_ARCHITECTURE_I386

    tags_hdr = multiboot2_tag_align()
    tags_hdr += multiboot2_tag_end()
    tags_size = len(tags_hdr)
    ....
 ```
 instead of the `add_tags` function. So when we would add a certain tag in the future we would create a `multiboot2_tag_<tag_name>(<potential_args>)`, where potential_args would be possible values that may change the behavior of the tag, e.g. `multiboot2_tag_reloc(min_addr, max_addr, align, preference)`.
 
Sorry, when I first made the suggestion, I thought we could like make a very generic class structure or something to generically instantiate tags, but now that I see them properly in the spec they can differ quite a log, some not even having the `flags` field... it may not be worth the future maintenance effort... I think it would be easier to just have a function that returns a tag for type of tag.
What do you think?

mariapana replied:
Yes, it would be better and cleaner this way. It could get complicated/ugly if complex tags are added down the road. I will update the implementation :smiley:

mogasergiu commented on support/scripts/multiboot2.py:
```diff
@@ -0,0 +1,245 @@
+    mb2_header_length = MULTIBOOT_HEADER_BASE_SIZE + tags_size
+    # We need to convert checksum to 4-byte signed integer range to avoid overflow
+    mb2_checksum = -(mb2_magic + mb2_architecture + mb2_header_length)
+    mb2_checksum = (mb2_checksum + 2 ** 32) % 2 ** 32
```
Maybe surround `2 ** 32` with paranthesis as well? Just to be sure.

mogasergiu commented on support/scripts/multiboot2.py:
```diff
@@ -0,0 +1,245 @@
+
+            new_shdr[0:ELF64_SHDR_LEN] = shdr[0:ELF64_SHDR_LEN]  # keep the rest
+            new_shdr[24:32] = self.prpnd_off(
+                shdr[24:32], mb2_header_end_off
```
keep these on the same line as the function call? just like you did for phdrs.

mogasergiu commented on support/scripts/multiboot2.py:
```diff
@@ -0,0 +1,245 @@
+        # The offset to the ELF64 Program/Sections Headers are at
+        # the end of the Multiboot Header and the end of the file respectively,
+        # so that we do not mess up the Multiboot2 header, which must exist
+        # in the first 8192 bytes. Although the specification does not enforce
```
Perhaps mention that the spec mandates that it is in the first 32768 bytes though. This seems to be a GRUB quirk only.

mogasergiu commented on support/scripts/multiboot2.py:
```diff
@@ -0,0 +1,245 @@
+        # Write ELF64 header with updated offsets
+        new_elf.write(elf64_ehdr)
+
+        # Append Multiboot2 header
```
> The Multiboot2 header must be contained completely within the first 32768 bytes of the OS image, and must be 64-bit aligned.

The same way `multiboot.py` checks for alignment
```python
        # Append the Multiboot header
        # NOTE: Multiboot header must be 4-byte aligned - luckily that is the
        # case with the size of the ELF32 header (40 bytes)
        if elf32.tell() & 0b11:
            raise Exception("ELF32 header is not a multiple of 4 bytes in size")
```
You should do the same here.

mogasergiu commented on plat/kvm/Linker.uk:
```diff
@@ -1,8 +1,7 @@
+ifeq ($(findstring y,$(CONFIG_KVM_BOOT_PROTO_MULTIBOOT) $(CONFIG_KVM_BOOT_PROTO_MULTIBOOT2)),y)
 KVM_LDFLAGS-y += -Wl,-m,elf_x86_64
 KVM_LDFLAGS-y += -Wl,--entry=_multiboot_entry
-ELF64_TO_32 = y
```
Since you removed this here, you should remove it below as well.

mogasergiu replied:
I think you misunderstood me. What I meant was having the following git commit history
- commit that removes this ELF64_TO_32 and replaces its if below with the MULTIBOOT config (mb2 support not in yet
- commit that introduces the script
- commit that adds multiboot2 support (C/asm changes and the addition to Linker.uk of multiboot2.py)
- commit that adds mb2 support and changes the function names in the mkukimg script.

mogasergiu commented on plat/kvm/Linker.uk:
```diff
@@ -81,6 +80,9 @@ $(KVM_IMAGE): $(KVM_IMAGE).dbg
 ifeq ($(ELF64_TO_32),y)
```
As in here :point_up_2: This should now be `CONFIG_KVM_BOOT_PROTO_MULTIBOOT` instead.
Which makes me realize I forgot about this ELF64_TO_32 thingy
Now I think the PR should have 4 commits:
- the commit that removes this ELF64_TO_32  and replaces its usecase with the check for `CONFIG_KVM_BOOT_PROTO_MULTIBOOT` instead
- ... the other 3 you have right now

mogasergiu commented on plat/kvm/include/kvm-x86/multiboot2.h:
```diff
@@ -0,0 +1,381 @@
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL ANY
+ *  DEVELOPER OR DISTRIBUTOR BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
```
Misalignment?

mogasergiu commented on plat/kvm/include/kvm-x86/multiboot2.h:
```diff
@@ -0,0 +1,381 @@
+};
+
+#endif /*  ! __ASSEMBLY__ */
+
```
Surplus newline

mogasergiu commented on plat/kvm/x86/multiboot.S:
```diff
@@ -38,7 +38,23 @@
+#elif CONFIG_KVM_BOOT_PROTO_MULTIBOOT2
+#include <kvm-x86/multiboot2.h>
+#else /* !CONFIG_KVM_BOOT_PROTO_MULTIBOOT &&
+	!CONFIG_KVM_BOOT_PROTO_MULTIBOOT2 */
```
Align the `!`

mogasergiu commented on plat/kvm/x86/multiboot.S:
```diff
@@ -38,7 +38,23 @@
+#else /* !CONFIG_KVM_BOOT_PROTO_MULTIBOOT &&
+	!CONFIG_KVM_BOOT_PROTO_MULTIBOOT2 */
+#error "Unknown multiboot version"
+#endif
```
Repeat the
```C
/* !CONFIG_KVM_BOOT_PROTO_MULTIBOOT &&
	!CONFIG_KVM_BOOT_PROTO_MULTIBOOT2 */
```

mogasergiu commented on plat/kvm/x86/multiboot.S:
```diff
@@ -38,7 +38,23 @@
+#else /* !CONFIG_KVM_BOOT_PROTO_MULTIBOOT &&
+	!CONFIG_KVM_BOOT_PROTO_MULTIBOOT2 */
+#error "Unknown multiboot version"
+#endif
```
Ditto

mogasergiu commented on plat/kvm/x86/multiboot.S:
```diff
@@ -49,7 +65,14 @@ lcpu_bootstack:
+	!CONFIG_KVM_BOOT_PROTO_MULTIBOOT2 */
+#error "Unknown multiboot version"
+#endif
 	ur_data	quad, lcpu_bootstack, 8
```
Ditto

mogasergiu commented on plat/kvm/x86/multiboot.S:
```diff
@@ -63,8 +86,7 @@ lcpu_boot_startup_args:
 ENTRY(_multiboot_entry)
-	/* only multiboot is supported for now */
-	cmpl	$MULTIBOOT_BOOTLOADER_MAGIC, %eax
+	cmpl $MULTIBOOT_MAGIC, %eax
```
Tab between instruction and first operand

mogasergiu commented on plat/kvm/x86/multiboot2.c:
```diff
@@ -0,0 +1,180 @@
+ */
+void multiboot2_entry(struct lcpu *lcpu, unsigned long addr)
+{
+	struct ukplat_bootinfo *bi;
```
I have a personal preference when it comes to variable declaration. I will only mention it once, feel free to disregard/apply it if you also dislike/like it. I generally declare these in reverse xmas tree order...
e.g. https://github.com/unikraft/unikraft/blob/staging/plat/kvm/efi.c#L299
Or pretty much how that whole file does it. There are many other files that follow these patterns. Though this is not mandated by our coding style (though I hope it will) and others do not really use this pattern, I personally like it better.

mogasergiu commented on plat/kvm/x86/multiboot2.c:
```diff
@@ -0,0 +1,180 @@
+	 * Initialize the first tag address, skipping total_size(4) and
+	 * reserved (4) fields of the boot information structure
+	 */
+	struct multiboot_tag *tag = (struct multiboot_tag *)(addr + 8);
```
Try to declare things at the beginning of the function. Keep the styling consistent with the codebase. That being said you could make this assignment in the for loop.

mogasergiu commented on plat/kvm/x86/multiboot2.c:
```diff
@@ -0,0 +1,180 @@
+#define multiboot2_crash(msg, rc)	ukplat_crash()
+
+/* Round up 'x' to the nearest multiple of 'MULTIBOOT_TAG_ALIGN' */
+#define ROUND_UP_TO_ALIGNMENT(x) \
```
For this kind of alignment macro's we use names like `ALIGN_UP` so I would rename this to `MULTIBOOT_TAG_ALIGN_UP` and have it be defined as
```C
#define MULTIBOOT_TAG_ALIGN_UP(x)                      ALIGN_UP((x), MULTIBOOT_TAG_ALIGN)
```
This way you use the already existing macro.

mogasergiu commented on plat/kvm/x86/multiboot2.c:
```diff
@@ -0,0 +1,180 @@
+	/* Parse Multiboot2 tags */
+	for (; tag->type != MULTIBOOT_TAG_TYPE_END;
+	     tag = (struct multiboot_tag *)((multiboot_uint8_t *)tag
+					+ ROUND_UP_TO_ALIGNMENT(tag->size))) {
```
Yeah this will look super ugly as a for loop given how long these definitions are... maybe let's do it as a while loop instead? :)

mogasergiu commented on plat/kvm/x86/multiboot2.c:
```diff
@@ -0,0 +1,180 @@
+			mrd.type = UKPLAT_MEMRT_INITRD;
+			mrd.flags = UKPLAT_MEMRF_READ;
+
+			#ifdef CONFIG_UKPLAT_MEMRNAME
```
Same commen as last time about not indenting `#ifdef`'s and using `#if`s instead. The `strncoy` looks overindented as well.

mogasergiu commented on plat/kvm/x86/multiboot2.c:
```diff
@@ -0,0 +1,180 @@
+			mrd_insert(bi, &mrd);
+			break;
+		case MULTIBOOT_TAG_TYPE_MMAP:
+			struct multiboot_tag_mmap *mmap_tag;
```
No late declarations, do all declarations at the beginning. And this switch case is kind of fat, I would move it to its own function. This way you get to keep your declarations :P

mogasergiu commented on support/scripts/mkukimg:
```diff
@@ -73,6 +73,41 @@ EOF
 	fi
 }
 
+mkukmb2iso()
```
All I have to say for this commit is to just name it `support/scripts: Support building multiboot2 images` as far as its commit subject goes. The fact that you did this minor rename for the old multiboot option is a small detail that can be mentioned in the main commit message only as you have already done.

mariapana committed:
plat/kvm: Remove ELF64_TO_32 for the Multiboot protocol

This commit removes the `ELF64_TO_32` configuration used for the Multiboot
protocol. As an alternative, its usecase is replaced with checks for
`CONFIG_KVM_BOOT_PROTO_MULTIBOOT`.

Signed-off-by: Maria Pana <maria.pana4@gmail.com>

mogasergiu commented on support/scripts/multiboot2.py:
```diff
@@ -0,0 +1,247 @@
+from struct import unpack
+
+# Default global values
+ELF_MAGIC = {
```
I don't see this being used anywhere, so maybe let's remove it.

mogasergiu commented on support/scripts/multiboot2.py:
```diff
@@ -0,0 +1,247 @@
+}
+
+ELF64_EHDR_LEN = 64
+
```
Can you remove these new lines between these declarations?

mogasergiu commented on support/scripts/multiboot2.py:
```diff
@@ -0,0 +1,247 @@
+
+MULTIBOOT_HEADER_TAG_OPTIONAL = 1
+
+MULTIBOOT_HEADER_TAG_END = 0
```
Let's only declare the tags that are used. And for those that are used give a brief description/comment for what they are used for and if there is any special treatment from GRUB2 worth mentioning (that is not mandated by the spec).

mogasergiu commented on support/scripts/multiboot2.py:
```diff
@@ -0,0 +1,247 @@
+
+
+def multiboot2_tag_end():
+    size = 8
```
Maybe let's provide an endianness argument to these tag function.
So like let's maybe go with
```python
def multiboot2_tag_module_align(endian):
    return (
        MULTIBOOT_HEADER_TAG_MODULE_ALIGN.to_bytes(2, endian)
        + MULTIBOOT_HEADER_TAG_OPTIONAL.to_bytes(2, endian)
        + (8).to_bytes(4, endian)
    )
```
Ditto for `get_multiboot2_hdr` and let's let the caller get the endianness from an ELF editor getter like `elf64.get_endian()`.

mogasergiu commented on plat/kvm/include/kvm-x86/multiboot2.h:
```diff
@@ -0,0 +1,380 @@
+#ifndef __MULTIBOOT2_H__
+#define __MULTIBOOT2_H__
+
+/*  How many bytes from the start of the file we search for the header. */
```
I see all comments have a two whitespace after `/*`. I guess this is from the original file. Let's remove them to comply with our coding style.

mogasergiu commented on plat/kvm/x86/multiboot2.c:
```diff
@@ -0,0 +1,188 @@
+void handle_mmap_tag(struct ukplat_memregion_desc mrd,
+		     struct ukplat_bootinfo *bi, struct multiboot_tag *tag)
+{
+	__sz i, mmap_total_sz, mmap_entry_cnt;
```
:muscle: :christmas_tree:

mariapana replied:
You mean for the function parameters as well or have I misunderstood how it should be done for the declarations starting from line 44?

mogasergiu replied:
Ah no lol just appreciating that you adopted the style :laughing:

mariapana replied:
Yeah, it's much better this way. Looking back on it, I don't know how I didn't do it like this before :)

mogasergiu replied:
Eh, it's just a matter of style :smile:. Unless mandated by the coding style policy, feel free to do things your own way. I just like this style visually. Got it from the Linux Kernel as they have started adopting this on the mailing lists as well.

mogasergiu commented on plat/kvm/x86/multiboot2.c:
```diff
@@ -0,0 +1,188 @@
+	while (tag->type != MULTIBOOT_TAG_TYPE_END) {
+		switch (tag->type) {
+		case MULTIBOOT_TAG_TYPE_CMDLINE:
+			struct multiboot_tag_string *tag_str;
```
Can you also declare this at the beginning of the function?

mogasergiu commented on plat/kvm/x86/multiboot2.c:
```diff
@@ -0,0 +1,188 @@
+			cmdline_len = tag_str->size;
+			break;
+		case MULTIBOOT_TAG_TYPE_MODULE:
+			struct multiboot_tag_module *mod_tag;
```
Ditto

mogasergiu commented on plat/kvm/x86/multiboot2.c:
```diff
@@ -0,0 +1,188 @@
+		multiboot2_crash("Cannot insert bootinfo memory region", rc);
+}
+
+void handle_mmap_tag(struct ukplat_memregion_desc mrd,
```
What is the `mrd` argument for? I think you may as well declare it on this function's frame.

mogasergiu commented on plat/kvm/x86/multiboot2.c:
```diff
@@ -0,0 +1,188 @@
+		if (m->type == MULTIBOOT_MEMORY_AVAILABLE) {
+			mrd.type  = UKPLAT_MEMRT_FREE;
+			mrd.flags = UKPLAT_MEMRF_READ |
+					UKPLAT_MEMRF_WRITE;
```
Misaligned with `=`

mogasergiu commented on plat/kvm/Linker.uk:
```diff
@@ -103,4 +102,4 @@ UK_IMAGES-$(CONFIG_OPTIMIZE_COMPRESS) += $(KVM_IMAGE).gz
 
 # ...for cleaning:
-LIBKVMPLAT_CLEAN += $(call build_clean,$(KVM_DEBUG_IMAGE).bootinfo)
+LIBKVMPLAT_CLEAN += $(call build_clean,$(KVM_DEBUG_IMAGE).bootinfo)
```
Pretty sure this should not appear in the diff :)

mogasergiu commented on support/scripts/multiboot2.py:
```diff
@@ -0,0 +1,241 @@
+
+    def get_endian(self):
+        return self.endian
+
```
Double newlines?
(Ditto for below)

mogasergiu commented on support/scripts/multiboot2.py:
```diff
@@ -0,0 +1,241 @@
+    return (
+        MULTIBOOT_HEADER_TAG_END.to_bytes(2, endian)
+        + MULTIBOOT_HEADER_TAG_OPTIONAL.to_bytes(2, endian)
+        + size.to_bytes(4, endian)
```
Why not just `(8).to_bytes(4, endian)`?
Also just noticed now, but operand should typically come at the end of the line you are continuing on a next line. So I was expecitng something like
```python
return (MULTIBOOT_HEADER_TAG_END.to_bytes(2, endian) + 
        MULTIBOOT_HEADER_TAG_OPTIONAL.to_bytes(2, endian) +
        (8).to_bytes(4, endian)
```

mogasergiu commented on plat/kvm/include/kvm-x86/multiboot2.h:
```diff
@@ -0,0 +1,380 @@
+#define MULTIBOOT_INFO_ALIGN					0x00000008
+
+/*  Flags set in the ’flags’ member of the multiboot header. */
+
```
Why the newline?

mogasergiu commented on plat/kvm/include/kvm-x86/multiboot2.h:
```diff
@@ -0,0 +1,380 @@
+#define MULTIBOOT_CONSOLE_FLAGS_CONSOLE_REQUIRED		1
+#define MULTIBOOT_CONSOLE_FLAGS_EGA_TEXT_SUPPORTED		2
+
+#ifndef __ASSEMBLY__
```
`#if !__ASSEMBLY__`

mogasergiu commented on plat/kvm/include/kvm-x86/multiboot2.h:
```diff
@@ -0,0 +1,380 @@
+	multiboot_uint32_t load_base_addr;
+};
+
+#endif /*  ! __ASSEMBLY__ */
```
Why the two whitespaces before the `!`? :thinking:

mogasergiu commented on plat/kvm/x86/multiboot2.c:
```diff
@@ -0,0 +1,185 @@
+	mmap = (struct multiboot_mmap_entry *)mmap_tag->entries;
+
+	mmap_total_sz = (mmap_tag->size -
+			 sizeof(struct multiboot_tag_mmap));
```
Maybe `mmap_total_sz = (mmap_tag->size - sizeof(*mmap_tag))`? Typically it is better to use the actual variable whose type size you are trying to find out. This way, on type rename, you need to update less lines (also shorter).

mogasergiu commented on plat/kvm/x86/multiboot2.c:
```diff
@@ -0,0 +1,185 @@
+		}
+
+		tag = (struct multiboot_tag *)((multiboot_uint8_t *)tag
+					+ MULTIBOOT_TAG_ALIGN_UP(tag->size));
```
Misaligned? Also `+` on the first line.

mogasergiu commented on plat/kvm/x86/multiboot2.c:
```diff
@@ -0,0 +1,185 @@
+	ukplat_memregion_list_coalesce(&bi->mrds);
+
+	/* Create cmdline mrd after coalescing */
+	char *cmdline = ukplat_memregion_alloc(cmdline_len,
```
Late declaration :)

mogasergiu reviewed (changes requested):
Some more unimportant nitpicks. Anyway, you will need to rebase now that some related PR's (the one I told you about for the cmdline issue) have been merged.
